[shawnnygoh] iP

shawnnygoh opened the pull request:
# Ally

> "The key is not to prioritize what's on your schedule, but to schedule your priorities." – Stephen Covey ([source](https://www.goodreads.com/quotes/706652-the-key-is-not-to-prioritize-what-s-on-your-schedule))

Ally is the future of task management. It offers:

* A **text-based** interface
* A **beginner friendly** system
* ~FAST~ _BLAZINGLY_ FAST performance

All you need to do is:

1. Download it from [here](https://github.com/shawnnygoh/ip).
2. Execute the file.
3. Add your tasks.
4. Let Ally manage your tasks for you 😉

And it's completely **FREE**!

## Features

- [x] Managing tasks
- [x] Managing deadlines 
- [ ] Reminders (coming soon)

If you're a Java programmer, you can also use Ally to practice too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

shawnnygoh committed:
Rename Duke to Ally and implement greeting

shawnnygoh committed:
Refactor Ally with OOP and add Echo functionality

shawnnygoh committed:
Add ToDoList class with add and list functionality

shawnnygoh committed:
Add Task class with mark as done or undone functionality

shawnnygoh committed:
Add ToDo, Deadline and Event class

shawnnygoh committed:
Add automated text UI testing

shawnnygoh committed:
Add error handling

shawnnygoh committed:
Add delete functionality

shawnnygoh committed:
Add Storage class with save tasks functionality

shawnnygoh committed:
Merge branch 'branch-Level-7'

- Add Storage class
- Add save tasks functionality

shawnnygoh committed:
Add DateTimeParser class with time interpretation functionality

shawnnygoh committed:
Merge pull request #1 from shawnnygoh/branch-Level-8

Add DateTimeParser class with time interpretation functionality

shawnnygoh committed:
Refactor code to follow more closely with OOP

shawnnygoh committed:
Merge branch 'branch-A-MoreOOP'

- Refactor code to follow more closely with OOP
- Create new Command, Parser, TaskList and AllyException classes

shawnnygoh committed:
Organize classes into suitable packages

shawnnygoh committed:
Merge branch 'branch-A-Packages'

- Organize classes into suitable packages

shawnnygoh committed:
Merge branch 'add-gradle-support'

- Add Gradle support to automate project build

shawnnygoh committed:
Add Gradle to automate project build

shawnnygoh committed:
Merge branch 'branch-A-Gradle'

- Add Gradle to automate project build

shawnnygoh committed:
Add JUnit tests for Todo and Deadline classes

shawnnygoh committed:
Merge branch 'branch-A-JUnit'

- Add JUnit tests for Todo and Deadline classes

shawnnygoh committed:
Enable Ally to be packaged into a JAR file

shawnnygoh committed:
Merge branch 'branch-A-Jar'

- Enable Ally to be packaged into a JAR file

shawnnygoh committed:
Add JavaDoc comments to most classes and public methods

shawnnygoh committed:
Merge branch 'branch-A-JavaDoc'

- Add JavaDoc comments to most classes and public methods

shawnnygoh committed:
Modify FindCommand to search by keyword

shawnnygoh committed:
Merge branch 'branch-Level-9'

- Modify FindCommand to search by keyword

shawnnygoh committed:
Add Checkstyle to detect coding style violations

shawnnygoh committed:
Merge branch 'branch-A-CheckStyle'

- Add Checkstyle to detect coding style violations

shawnnygoh committed:
Add GUI with JavaFX

shawnnygoh committed:
Merge branch 'branch-Level-10'

- Add GUI with JavaFX

swatirajesh277 commented on src/main/java/ally/allyexception/InvalidTaskNumberException.java:
```diff
@@ -0,0 +1,13 @@
+     * Constructs an InvalidTaskNumberException.
+     */
+    public InvalidTaskNumberException() {
+        super("Sorry! The task number is invalid. 😵");
```
I like the use of emoji here! Very unique!

swatirajesh277 commented on src/main/java/ally/command/AddEventCommand.java:
```diff
@@ -0,0 +1,81 @@
+    private final LocalDateTime startTime;
+    private final LocalDateTime endTime;
+
+    private String response = "Command executed";
```
Intuitive string name! Nice!

swatirajesh277 commented on src/main/java/ally/tasklist/Task.java:
```diff
@@ -0,0 +1,75 @@
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
I like how you followed the java naming convention for boolean variables

swatirajesh277 commented on src/main/java/ally/ui/UI.java:
```diff
@@ -0,0 +1,192 @@
+        System.out.println(LOGO);
+        System.out.println("Hi! I'm Ally!");
+        System.out.println("What can I do for you?");
+        showDivider();
```
Good use of having divider as a method! I like this a lot

swatirajesh277 commented on src/main/java/ally/parser/Parser.java:
```diff
@@ -0,0 +1,189 @@
+     * @param timeStr String to check
+     * @return true if string is time-only format
+     */
+    public static boolean isTimeOnly(String timeStr) {
```
Intuitive method name! You have followed convention very well

swatirajesh277 reviewed (commented):
Wow your code is really amazing and I learnt a lot from it. I could hardly find any nits and that is very impressive! Keep up the good work!

shawnnygoh committed:
Add assertions throughout codebase

Added assertions to verify:
* Valid input format for parseCommand method in Parser
* Task indices within bounds in TaskList
* Task data has the required 3 parts when parsing from storage.

shawnnygoh committed:
Refactor code to improve code quality

Split parseDayFormat into two methods to shorten method length for better readability.

shawnnygoh committed:
Merge pull request #2 from shawnnygoh/branch-A-Assertions

Add assertions

shawnnygoh committed:
Merge branch 'master' into branch-A-CodeQuality

shawnnygoh committed:
Merge pull request #3 from shawnnygoh/branch-A-CodeQuality

Refactor code to improve code quality

shawnnygoh committed:
Add GitHub Actions CI workflow

shawnnygoh committed:
Merge pull request #4 from shawnnygoh/branch-A-CI

Add GitHub Actions CI workflow

shawnnygoh committed:
Add GitHub Actions CI workflow

shawnnygoh committed:
Merge branch 'branch-A-CI'

shawnnygoh committed:
Merge branch 'master' of github.com:shawnnygoh/ip

shawnnygoh committed:
Add GitHub Actions CI workflow

shawnnygoh committed:
Merge branch 'branch-A-CI'

shawnnygoh committed:
Add Help Command to provide in-app guidance to users

shawnnygoh committed:
Merge branch 'branch-C-Help'

shawnnygoh committed:
Fix GitHub Actions CI workflow

shawnnygoh committed:
Merge branch 'branch-A-CI'

Masunori commented on src/main/java/ally/allyexception/AllyException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * Custom exception class for handling specific exceptions in the Ally application.
+ */
+public class AllyException extends Exception {
```
I like how you structure all your exceptions under AllyException. It helps to debug easier as you know if it is caused by the programmer or the user's wrong syntax! 👍

Masunori commented on src/main/java/ally/allyexception/EmptyTaskNameException.java:
```diff
@@ -0,0 +1,15 @@
+     * Constructs an EmptyTaskNameException with a specific task type.
+     *
+     * @param taskType the type of the task (e.g., todo, deadline, or event).
+     */
```
For all errors related to the user's syntax error, I think that you can add another line to show them the correct thing to do. For this one as example, the message can be like:

`Sorry! The name of a to do cannot be empty.`
`Please specify the to do in the format: to do [task_description]!`

With this, the user would immediately know the correct thing to do without needing to open the help section 😄

Masunori commented on src/main/java/ally/command/Command.java:
```diff
@@ -0,0 +1,35 @@
+
+/**
+ * Abstract base class for all commands.
+ */
```
Very structured!

Masunori commented on src/main/java/ally/command/HelpCommand.java:
```diff
@@ -0,0 +1,36 @@
+
+/**
+ * Command to access help page.
+ */
```
Very helpful command!

Masunori commented on src/main/java/ally/command/Command.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Abstract base class for all commands.
+ */
+public abstract class Command {
```
I notice that all your subclasses of Command have to have their own String response. How about putting a 
`protected String response` in this class instead?

Masunori commented on src/main/java/ally/allyexception/InvalidTaskNumberException.java:
```diff
@@ -0,0 +1,13 @@
+    /**
+     * Constructs an InvalidTaskNumberException.
+     */
+    public InvalidTaskNumberException() {
```
When this exception is thrown, how about you inform the user of the range of numbers they can key in? As a user, I would find this helpful!

Masunori commented on src/main/resources/css/dialog-box.css:
```diff
@@ -0,0 +1,58 @@
+.label {
```
A lot of effort! Ten out of ten!

Masunori commented on src/main/java/ally/gui/DialogBox.java:
```diff
@@ -0,0 +1,101 @@
+        db.changeDialogStyle(commandType);
+        return db;
+    }
+
```
Your approach is completely fine, but I just want to share my perspective, so that both of us benefit. My approach for this is that I use a HashMap that maps from the command string to a function, in this case, it will look like this:

```java
new HashMap<String, Function<Void, Void>> dummyNameHashMap;
```

and then your block of execution code will become 
```java
() -> dialog.getStyleClass().add('some-class')
```
then your changeDialogStyle will be:
```java
private void changeDialogStyle(String commandType) {
    this.dummyNameHashMap.get(commandType).apply();
}
```

Masunori commented on src/main/java/ally/parser/Parser.java:
```diff
@@ -0,0 +1,197 @@
+            return new ListCommand();
+        case "bye":
+            return new ExitCommand();
+        case "find":
```
How about your FindCommand() handling the error instead?

Masunori commented on src/main/java/ally/parser/Parser.java:
```diff
@@ -0,0 +1,197 @@
+     * @param dateStr String containing date/time
+     * @return LocalDateTime object
+     * @throws DateTimeParseException if string cannot be parsed
+     */
```
If you use this outside your Parser and I have missed it, you can ignore this comment.

As this is a utility method within Parser and logically I would not ask something outside a Parser to parse, how about we make this private?

Masunori commented on src/main/java/ally/storage/Storage.java:
```diff
@@ -0,0 +1,136 @@
+
+        return tasks;
+    }
+
```
(CS2030S Tell, Don't Ask incoming 😨 ...)

How about you tell the Task to encode itself instead of asking for its fields and doing the encoding manually here?

Masunori commented on src/main/java/ally/storage/Storage.java:
```diff
@@ -0,0 +1,136 @@
+
+        return baseEncoding;
+    }
+
```
Same thing here, how about providing a utility method from the Task class that will help you do this decoding?

Masunori commented on src/main/java/ally/tasklist/Task.java:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * Represents a generic task with a description and completion status.
+ */
```
Your Task class overall is very structured and very OOP 👍

Masunori commented on src/main/java/ally/tasklist/TaskList.java:
```diff
@@ -0,0 +1,122 @@
+ * Manages a list of tasks, including adding, removing, and updating tasks.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
```
For my side, I have one method Storage::load(TaskList) as a single entry point of interaction. Then, Parser.java, upon receiving a Task, will invoke TaskList::add(Task) and Storage::saveTask(Task). In my opinion, these two are two standalone entities.

But of course, there is no one correct design, and I would love to hear about your rationale behind your design 😄. From what I see, you hide the Storage away from your Ally.java, and I am curious about this too.

Masunori commented on src/main/resources/images/ally.png:
Cute!

Masunori commented on text-ui-test/data/ally.txt:
You can .gitignore this entire folder

Masunori commented on src/test/java/ally/tasklist/DeadlineTest.java:
```diff
@@ -0,0 +1,41 @@
+import org.junit.jupiter.api.Test;
+
+public class DeadlineTest {
+
```
A little nit, the naming convention for test methods is:
```java
public void testFeature_testScenario_expectedBehavior();
```

For example,
```java
public void toString_deadlineTask_correctFormat();
```

If you do not have the second and third part, it may imply to a person from its name, that testToString of yours will test all possible scenarios. 

Other than that, the test looks comprehensive to me!

shawnnygoh committed:
Improve code quality and maintain coding standard

shawnnygoh committed:
Add app icon and title

shawnnygoh committed:
Add User Guide

shawnnygoh committed:
Merge branch 'branch-A-UserGuide'

shawnnygoh committed:
Add final tweaks
